RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -358,38 +529,258 @@ void TrackExtenderWithMTDT<TrackCollection>::produce(edm::Event& ev, const edm::
+
+      pathlength = length;
+
+      gammasq_pi = 1. + magp2 * m_pi_inv2;
+      beta_pi = std::sqrt(1. - 1. / gammasq_pi);
+      dt_pi = pathlength / beta_pi * c_inv;
+
+      gammasq_k = 1. + magp2 * m_k_inv2;
+      beta_k = std::sqrt(1. - 1. / gammasq_k);
+      dt_k = pathlength / beta_k * c_inv;
+
+      gammasq_p = 1. + magp2 * m_p_inv2;
+      beta_p = std::sqrt(1. - 1. / gammasq_p);
+      dt_p = pathlength / beta_p * c_inv;
+
+      dt = tmtd - dt_pi - t_vtx;  //assume by default the pi hypothesis
+      dterror = sqrt(tmtderror * tmtderror + t_vtx_err * t_vtx_err);
+      betaerror = 0;
+      if (addPIDError) {
+        dterror = sqrt(dterror * dterror + (dt_p - dt_pi) * (dt_p - dt_pi));
+        betaerror = beta_p - beta_pi;
+      }
+
+      dtchi2 = (dt * dt) / (dterror * dterror);
+
+      dt_best = dt;
+      dterror_best = dterror;
+      dtchi2_best = dtchi2;
+
+      prob_pi = -1.;
+      prob_k = -1.;
+      prob_p = -1.;
+
+      if (!addPIDError) {
+        //*TODO* deal with heavier nucleons and/or BSM case here?
+        double chi2_pi = dtchi2;
+        double chi2_k = (tmtd - dt_k - t_vtx) * (tmtd - dt_k - t_vtx) / (dterror * dterror);
+        double chi2_p = (tmtd - dt_p - t_vtx) * (tmtd - dt_p - t_vtx) / (dterror * dterror);
+
+        double rawprob_pi = exp(-0.5 * chi2_pi);
+        double rawprob_k = exp(-0.5 * chi2_k);
+        double rawprob_p = exp(-0.5 * chi2_p);
+        double normprob = 1. / (rawprob_pi + rawprob_k + rawprob_p);
+
+        prob_pi = rawprob_pi * normprob;
+        prob_k = rawprob_k * normprob;
+        prob_p = rawprob_p * normprob;
+
+        double prob_heavy = 1. - prob_pi;
+        constexpr double heavy_threshold = 0.75;
+
+        if (prob_heavy > heavy_threshold) {
+          if (chi2_k < chi2_p) {
+            dt_best = (tmtd - dt_k - t_vtx);
+            dtchi2_best = chi2_k;
+          } else {
+            dt_best = (tmtd - dt_p - t_vtx);
+            dtchi2_best = chi2_p;
+          }
+        }
+      }
+    }
+
+    static constexpr double m_pi = 0.13957018;
+    static constexpr double m_pi_inv2 = 1.0 / m_pi / m_pi;
+    static constexpr double m_k = 0.493677;
+    static constexpr double m_k_inv2 = 1.0 / m_k / m_k;
+    static constexpr double m_p = 0.9382720813;
+    static constexpr double m_p_inv2 = 1.0 / m_p / m_p;
+    static constexpr double c_cm_ns = CLHEP::c_light * CLHEP::ns / CLHEP::cm;  //[cm/ns]
+    static constexpr double c_inv = 1.0 / c_cm_ns;
+    //this should come from BeamSpot in principle
+    //static constexpr double bserror = 0.18;
+
+    double tmtd;
+    double tmtderror;
+
+    double pathlength;
+
+    double dt;
+    double dterror;
+    double dtchi2;
+
+    double dt_best;
+    double dterror_best;
+    double dtchi2_best;
+
+    double gammasq_pi;
+    double beta_pi;
+    double dt_pi;
+
+    double gammasq_k;
+    double beta_k;
+    double dt_k;
+
+    double gammasq_p;
+    double beta_p;
+    double dt_p;
+
+    double prob_pi;
```

Comment:
Seems like these can be grouped as well; esp in case some other option will show up (like "D" or alpha).
```gamasq, prob, beta, dt``` can be combined in a struct; then pi, k, p can be in an array